server/src/main/java/org/elasticsearch/cluster/coordination/JoinValidationService.java

```diff
@@ -0,0 +1,366 @@
+        private final DiscoveryNode discoveryNode;
+
+        JoinValidation(DiscoveryNode discoveryNode, ActionListener<TransportResponse.Empty> listener) {
+            super(listener);
+            this.discoveryNode = discoveryNode;
+        }
+
+        @Override
+        protected void doRun() throws Exception {
+            assert discoveryNode.getVersion().onOrAfter(Version.V_8_3_0) : discoveryNode.getVersion();
+            // NB these things never run concurrently to each other, or to the cache cleaner (see IMPLEMENTATION NOTES above) so it is safe
+            // to do these (non-atomic) things to the (unsynchronized) statesByVersion map.
+            final var cachedBytes = statesByVersion.get(discoveryNode.getVersion());
+            final var bytes = Objects.requireNonNullElseGet(cachedBytes, () -> serializeClusterState(discoveryNode));
+            assert bytes.hasReferences() : "already closed";
+            bytes.incRef();
+            transportService.sendRequest(
+                discoveryNode,
+                JOIN_VALIDATE_ACTION_NAME,
+                new BytesTransportRequest(bytes, discoveryNode.getVersion()),
+                REQUEST_OPTIONS,
+                new ActionListenerResponseHandler<>(
+                    ActionListener.runAfter(listener, bytes::decRef),
+                    in -> TransportResponse.Empty.INSTANCE,
+                    ThreadPool.Names.CLUSTER_COORDINATION
+                )
+            );
+            if (cachedBytes == null) {
+                transportService.getThreadPool().schedule(new Runnable() {
+                    @Override
+                    public void run() {
+                        execute(cacheClearer);
+                    }
+
+                    @Override
+                    public String toString() {
+                        return cacheClearer + " after timeout";
+                    }
+                }, cacheTimeout, ThreadPool.Names.CLUSTER_COORDINATION);
+            }
+        }
+
+        @Override
+        public String toString() {
+            return "send cached join validation request to " + discoveryNode;
+        }
+    }
+
+    private ReleasableBytesReference serializeClusterState(DiscoveryNode discoveryNode) {
+        final var bytesStream = transportService.newNetworkBytesStream();
+        var success = false;
+        try {
+            final var clusterState = clusterStateSupplier.get();
+            final var version = discoveryNode.getVersion();
+            try (
+                var stream = new OutputStreamStreamOutput(
+                    CompressorFactory.COMPRESSOR.threadLocalOutputStream(Streams.flushOnCloseStream(bytesStream))
+                )
+            ) {
+                stream.setVersion(version);
+                clusterState.writeTo(stream);
+            } catch (IOException e) {
+                throw new ElasticsearchException("failed to serialize cluster state for publishing to node {}", e, discoveryNode);
```

Comment:
Mmm possibly, especially with an `IOException`, but in general exceptions during serialization are the last line of defence if there's some new feature that simply can't be represented in an older version. It's definitely better to avoid using new features in mixed clusters where possible but it's not a hard constraint.